File: code_wars/4kyu_permutation.py
# Problem:Permutation(4kyu)
"""
Instructions
"""
# My Solution:

import itertools
def permutations(str):
    return [''.join(i) for i in list(set(itertools.permutations(str)))]
# Other Solutions:

# Solution: By warriors plantpark

def permutations(s):        
    if(len(s)==1): return [s]
    result=[]
    for i,v in enumerate(s):
        result += [v+p for p in permutations(s[:i]+s[i+1:])]
    return list(set(result))

# Solution: By Edi4ka

def permutations(word):
    if len(word)<=1:
        return [word]

    perms=permutations(word[1:])
    char=word[0]
    result=[]
    for perm in perms:
        for i in range(len(perm)+1):
            result.append(perm[:i] + char + perm[i:])
    return set(result)

 # Solution: By   Sebek

def permutations(s):
    if len(s) == 0:
        return []
    elif len(s) == 1:
        return [s]
    else:
        return set(s[i]+p for i in range(len(s)) for p in permutations(s[:i] + s[i+1:]))

File: code_wars/JavaScript/4kyu_permutation.js

// Problem:Permutation 4kyu
/*
In this kata you have to create all permutations of an input string and remove duplicates, if present. This means, you have to shuffle all letters from the input in all possible orders.

Examples:

permutations('a'); // ['a']
permutations('ab'); // ['ab', 'ba']
permutations('aabb'); // ['aabb', 'abab', 'abba', 'baab', 'baba', 'bbaa']
The order of the permutations doesn't matter.
*/
// My Solution:

function permutations(string) {

  if(string.length <= 1){return [string];}

  let perms = permutations(string.slice(1, string.length));
  let char = string[0];
  let result = [];

  perms.forEach(perm => {

    for(let i = 0; i <= perm.length; i++){

      result.push(perm.slice(0, i) + char + perm.slice(i, perm.length));
    }
  })

  return Array.from(new Set(result));  
}
console.log(permutations('aabb'))

// Other Solutions:

// Solution: By warriors Azuaron, MinionKevin, testtest123, jwong483, Vorgue, sharpskill (plus 19 more warriors)

function permutations(string) {

  var arr = string.split(''), tmp = arr.slice(), heads = [], out = [];

  if(string.length == 1) return [string];

  arr.forEach(function(v, i, arr) {

    if(heads.indexOf(v) == -1) {

      heads.push(v);
      tmp.splice(tmp.indexOf(v), 1);

      permutations(tmp.join('')).forEach(function(w) {out.push(v + w);});
      
      tmp.push(v);
    }
  });
  return out;
}

// Solution: Bythink2011, melanoman, willin, dubdjon, July

function permutations(string) {

  return (string.length == 1) ? [string] : string.split('').map(

     (e, i) => permutations(string.slice(0,i) + string.slice(i+1)).map((e2) => e+e2)

  ).reduce((r,e) => r.concat(e)).sort().filter((e,i,a) => (i==0) || a[i-1] != e);
}

// Solution: By evk

function permutations(str) {

 return (str.length <= 1) ? [str] :

         Array.from(new Set(

           str.split('')

              .map((char, i) => permutations(str.substr(0, i) + str.substr(i + 1)).map(p => char + p))

              .reduce((r, x) => r.concat(x), [])
         ));
}